Allow WebRTC network traffic on the interface used to load the document main resource even if it is not the default interface

youennf opened the pull request:
#### b5877b229ffa2ae13da637a0fd8910e30ffafc29
<pre>
Allow WebRTC network traffic on the interface used to load the document main resource even if it is not the default interface
<a href="https://bugs.webkit.org/show_bug.cgi?id=275986">https://bugs.webkit.org/show_bug.cgi?id=275986</a>
<a href="https://rdar.apple.com/problem/130748643">rdar://problem/130748643</a>

Reviewed by Eric Carlson.

With split VPN, HTTP trafic may go through either default interface or the VPN interface.
But WebRTC UDP traffic will only go through the default interface as we do not enumerate all interfaces even post getUserMedia.

To improve the support of split VPN, we use the following heuristic:
- We continue allow using the default network interface.
- We get the interface that would be used to create a TCP connection to the document&apos;s host.

In case of no VPN, or full VPN, the same interface will be used.
In case of split VPN, two interfaces may be used.
This only happens post getUserMedia and is controlled by EnumeratingVisibleNetworkInterfacesEnabled flag.

* LayoutTests/http/wpt/webrtc/resources/cacheable-iframe.py: Added.
(main):
* LayoutTests/http/wpt/webrtc/rtcNetworkInterface-expected.txt: Added.
* LayoutTests/http/wpt/webrtc/rtcNetworkInterface.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/RTCNetworkManager.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::rtcNetworkInterfaceName const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.cpp:
(WebKit::NetworkRTCProvider::getInterfaceName):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.messages.in:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::createNWConnection):
(WebKit::NetworkRTCTCPSocketCocoa::NetworkRTCTCPSocketCocoa):
(WebKit::NetworkRTCTCPSocketCocoa::getInterfaceName):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::networksChanged):
(WebKit::LibWebRTCNetworkManager::interfaceNameForTesting const):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.h:

Canonical link: <a href="https://commits.webkit.org/280535@main">https://commits.webkit.org/280535@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/001d15f0a71843d42a96c5c8b51212919f6ce348

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36227 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9373 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60518 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43850 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7532 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46088 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5155 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58929 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34029 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49109 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26946 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30809 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6440 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6346 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6711 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62200 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/812 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6813 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53342 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/815 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49164 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53382 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12579 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/691 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32056 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33141 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34226 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32887 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1107,7 +1106,12 @@ - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)data
+        auto responseMetrics = WebCore::copyTimingData(taskMetrics, networkDataTask->networkLoadMetrics());
+        if (responseMetrics && networkDataTask->isMainResource()) {
+            networkDataTask->networkLoadMetrics().interfaceName = String(metrics._interfaceName);
+            responseMetrics->interfaceName = networkDataTask->networkLoadMetrics().interfaceName;
```
If there is a reason not to do something like the following, you should add a comment explaining it is necessary

```suggestion
            auto interfaceName = String(metrics._interfaceName);
            networkDataTask->networkLoadMetrics().interfaceName = interfaceName;
            responseMetrics->interfaceName = interfaceName;
```

youennf replied:
I removed this code, this was used to keep things consistent so that the metrics on the response and on end of load would be the same for interfaceName.

youennf committed:
Allow WebRTC network traffic on the interface used to load the document main resource even if it is not the default interface
https://bugs.webkit.org/show_bug.cgi?id=275986
rdar://problem/130748643

Reviewed by Eric Carlson.

With split VPN, HTTP trafic may go through either default interface or the VPN interface.
But WebRTC UDP traffic will only go through the default interface as we do not enumerate all interfaces even post getUserMedia.

To improve the support of split VPN, we use the following heuristic:
- We continue allow using the default network interface.
- We get the interface that would be used to create a TCP connection to the document's host.

In case of no VPN, or full VPN, the same interface will be used.
In case of split VPN, two interfaces may be used.
This only happens post getUserMedia and is controlled by EnumeratingVisibleNetworkInterfacesEnabled flag.

* LayoutTests/http/wpt/webrtc/resources/cacheable-iframe.py: Added.
(main):
* LayoutTests/http/wpt/webrtc/rtcNetworkInterface-expected.txt: Added.
* LayoutTests/http/wpt/webrtc/rtcNetworkInterface.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/RTCNetworkManager.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::rtcNetworkInterfaceName const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.cpp:
(WebKit::NetworkRTCProvider::getInterfaceName):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.messages.in:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::createNWConnection):
(WebKit::NetworkRTCTCPSocketCocoa::NetworkRTCTCPSocketCocoa):
(WebKit::NetworkRTCTCPSocketCocoa::getInterfaceName):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::networksChanged):
(WebKit::LibWebRTCNetworkManager::interfaceNameForTesting const):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.h:

Canonical link: https://commits.webkit.org/280535@main
